doc: Give examples of what DEVELHELP does

chrysn opened the pull request:
### Contribution description

This elaborates on the brief "additional code parts" line by listing kinds of changes made by DEVELHELP and giving some examples.

(On indentation: The current version built as an indented preformatted block in my doxygen version, so I unindented to make it testable for me).

### Extension

There's two stages to make this more comprehensive:

* Comprehensively list all *kinds* of things that depend on DEVELHELP. The list here may already do this, but I didn't go through every single DEVELHELP mention in the code base to verify that. New kinds of changes would need documenting then.

  Actually, this is what I'd prefer to do, so I'm marking the PR as a draft until I've done a full pass over `git grep DEVELHELP`.

* List every single change (possibly grouped by the above) -- this seems to be overkill. If doxygen can be convinced to have backlinks, it may be a good idea to make everything whose behavior depends on DEVELHELP link there (and thus be listed), though.

### Issues/PRs references

Suggested around https://github.com/RIOT-OS/RIOT/pull/14378#issuecomment-652639994

miri64 commented on doc.txt:
With `DEVELHELP=0`, [`NDEBUG` is enabled (unless `FORCE_ASSERTS=1`)](https://github.com/RIOT-OS/RIOT/blob/045874abe6c05d8ea3b340dfcd5882cea1bb07bf/makefiles/cflags.inc.mk#L56-L60). In return this means: usually with `DEVELHELP` (unless otherwise configured, e.g. with `NDEBUG` enabled) asserts are enabled.

chrysn replied:
Good point, thanks, added.

benpicco commented:
Why WIP?
This is already a good improvement over the status quo (no documentation), so even if we find that we have to extend the list, it would already help users better understand what the option does as is.

benpicco commented:
I'm afraid you'll have to force-push this one for the remaining CI tests to kick in.

chrysn commented:
Rebased to master for CI re-run. Removed WIP as right, this can still be improved (and it doesn't seem like I get to it quickly, thus this would just linger to rot).

chrysn commented:
With one ACK in, this is only blocked by check-labels that as I understand waits for miri's comments review to be resolved. The point on assert was added, so all should be addressed. @miri64, can you update the review status?

benpicco commented:
The label checker gets stuck on force pushes.
Removing / Adding a label gets it going again.
